File: aethoria-ui/src/types/monster.ts

// Path: src/types/monster.ts

export interface MonsterAbility {
    name: string;
    desc: string;
}

export interface MonsterAction extends MonsterAbility {
    // Actions can have more specific fields if needed in the future
    // e.g., attack_bonus, damage_dice, etc.
}

// This matches the `MonsterPublic` schema from the backend
export interface Monster {
    id: number;
    name: string;
    size: string;
    creature_type: string;
    alignment: string;
    armor_class: number;
    hit_dice: string;
    speed: { [key: string]: string };
    strength: number;
    dexterity: number;
    constitution: number;
    intelligence: number;
    wisdom: number;
    charisma: number;
    proficiencies: { value: number; proficiency: { name: string } }[] | null;
    damage_vulnerabilities: string[] | null;
    damage_resistances: string[] | null;
    damage_immunities: string[] | null;
    condition_immunities: string[] | null;
    senses: { [key: string]: string | number };
    languages: string;
    special_abilities: MonsterAbility[] | null;
    actions: MonsterAction[] | null;
    legendary_actions: MonsterAbility[] | null;
}

File: api/app/schemas/monster.py
# Path: api/app/schemas/monster.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any

class MonsterBase(BaseModel):
    name: str = Field(..., max_length=255)
    size: str = Field(..., max_length=50)
    creature_type: str = Field(..., max_length=100)
    alignment: str = Field(..., max_length=100)
    armor_class: int
    hit_points: int
    hit_dice: str = Field(..., max_length=50)
    speed: Dict[str, Any]
    
    strength: int
    dexterity: int
    constitution: int
    intelligence: int
    wisdom: int
    charisma: int

    proficiencies: Optional[List[Dict[str, Any]]] = None
    damage_vulnerabilities: Optional[List[str]] = None
    damage_resistances: Optional[List[str]] = None
    damage_immunities: Optional[List[str]] = None
    condition_immunities: Optional[List[str]] = None
    senses: Dict[str, Any]
    languages: str
    challenge_rating: float
    xp: int

    special_abilities: Optional[List[Dict[str, Any]]] = None
    actions: Optional[List[Dict[str, Any]]] = None
    legendary_actions: Optional[List[Dict[str, Any]]] = None

class MonsterCreate(MonsterBase):
    pass

class Monster(MonsterBase):
    id: int
    
    class Config:
        from_attributes = True

# --- NEW: Public-facing Monster Schema ---
# This schema omits sensitive DM-only information
class MonsterPublic(BaseModel):
    id: int
    name: str
    size: str
    creature_type: str
    alignment: str
    armor_class: int
    hit_dice: str # Show the dice, but not the calculated total HP
    speed: Dict[str, Any]
    
    strength: int
    dexterity: int
    constitution: int
    intelligence: int
    wisdom: int
    charisma: int

    proficiencies: Optional[List[Dict[str, Any]]] = None
    damage_vulnerabilities: Optional[List[str]] = None
    damage_resistances: Optional[List[str]] = None
    damage_immunities: Optional[List[str]] = None
    condition_immunities: Optional[List[str]] = None
    senses: Dict[str, Any]
    languages: str

    special_abilities: Optional[List[Dict[str, Any]]] = None
    actions: Optional[List[Dict[str, Any]]] = None
    legendary_actions: Optional[List[Dict[str, Any]]] = None
    
    class Config:
        from_attributes = True
# --- END NEW ---


